# Challenge 1 - T-test

In statistics, t-test is used to test if two data samples have a significant difference between their means. There are two types of t-test:

* **Student's t-test** (a.k.a. independent or uncorrelated t-test). This type of t-test is to compare the samples of **two independent populations** (e.g. test scores of students in two different classes). `scipy` provides the [`ttest_ind`](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ttest_ind.html) method to conduct student's t-test.

* **Paired t-test** (a.k.a. dependent or correlated t-test). This type of t-test is to compare the samples of **the same population** (e.g. scores of different tests of students in the same class). `scipy` provides the [`ttest_re`](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ttest_rel.html) method to conduct paired t-test.

Both types of t-tests return a number which is called the **p-value**. If p-value is below 0.05, we can confidently declare the null-hypothesis is rejected and the difference is significant. If p-value is between 0.05 and 0.1, we may also declare the null-hypothesis is rejected but we are not highly confident. If p-value is above 0.1 we do not reject the null-hypothesis.

Read more about the t-test in [this article](https://researchbasics.education.uconn.edu/t-test/) and [this Quora](https://www.quora.com/What-is-the-difference-between-a-paired-and-unpaired-t-test). Make sure you understand when to use which type of t-test. 

In [1]:
# Import libraries

import pandas as pd

#### Import dataset

In this challenge we will work on the Pokemon dataset you have used last week. The goal is to test whether different groups of pokemon (e.g. Legendary vs Normal, Generation 1 vs 2, single-type vs dual-type) have different stats (e.g. HP, Attack, Defense, etc.).

In [5]:
# Import dataset

pokemon = pd.read_csv('../data/Pokemon.csv')

pokemon.head()

,#,Name,Type 1,Type 2,Total,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,318,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,405,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,525,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,625,80,100,123,122,120,80,1,False
4,4,Charmander,Fire,NaN,309,39,52,43,60,50,65,1,False


#### First we want to define a function with which we can test the means of a feature set of two samples. 

In the next cell you'll see the annotations of the Python function that explains what this function does and its arguments and returned value. This type of annotation is called **docstring** which is a convention used among Python developers. The docstring convention allows developers to write consistent tech documentations for their codes so that others can read. It also allows some websites to automatically parse the docstrings and display user-friendly documentations.

Follow the specifications of the docstring and complete the function.

In [6]:
def t_test_features(s1, s2, features=['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Total']):
    """Test means of a feature set of two samples
    
    Args:
        s1 (dataframe): sample 1
        s2 (dataframe): sample 2
        features (list): an array of features to test
    
    Returns:
        dict: a dictionary of t-test scores for each feature where the feature name is the key and the p-value is the value
    """
    from scipy.stats import ttest_ind
    results = {}

    for feature in features:
        # Perform an independent t-test for the current feature
        # We assume equal_var=False as a safe default (Welch's t-test)
        t_stat, p_value = ttest_ind(s1[feature], s2[feature], equal_var=False)
        
        # Store the p-value in the results dictionary
        results[feature] = p_value
    
    return results

#### Using the `t_test_features` function, conduct t-test for Lengendary vs non-Legendary pokemons.

*Hint: your output should look like below:*

```
{'HP': 1.0026911708035284e-13,
 'Attack': 2.520372449236646e-16,
 'Defense': 4.8269984949193316e-11,
 'Sp. Atk': 1.5514614112239812e-21,
 'Sp. Def': 2.2949327864052826e-15,
 'Speed': 1.049016311882451e-18,
 'Total': 9.357954335957446e-47}
 ```

In [9]:
# Your code here
# Divide the dataset into Legendary and Non-Legendary groups
legendary_pokemon = pokemon[pokemon['Legendary'] == True]
non_legendary_pokemon = pokemon[pokemon['Legendary'] == False]

#Conduct the t-test for all features using function t_test_features
pokemon_ttest_results = t_test_features(legendary_pokemon, non_legendary_pokemon)

print("Legendary Pokemon vs. Non-Legendary Pokemon t-test results:")
for feature, p_value in pokemon_ttest_results.items():
    print(f"{feature:10}: {p_value}")

Legendary Pokemon vs. Non-Legendary Pokemon t-test results:
HP        : 1.002691170803531e-13
Attack    : 2.52037244923666e-16
Defense   : 4.82699849491934e-11
Sp. Atk   : 1.551461411223974e-21
Sp. Def   : 2.2949327864052925e-15
Speed     : 1.0490163118824546e-18
Total     : 9.357954335957505e-47


#### From the test results above, what conclusion can you make? Do Legendary and non-Legendary pokemons have significantly different stats on each feature?

In [ ]:
# Your comment here


### Conclusion: Legendary vs. Non-Legendary Pokémon Analysis

Based on the independent t-test results, we can draw the following statistical conclusions:

#### 1. Statistical Significance
**Yes**, there is a **statistically significant difference** between Legendary and non-Legendary Pokémon across every single feature. 

* **Evidence:** All calculated p-values are extremely close to zero (ranging from $10^{-13}$ to $10^{-47}$).
* **Decision:** Since every **p-value < 0.05**, we strictly **reject the null hypothesis ($H_0$)** for all categories.

#### 2. Key Findings
* **Objective Superiority:** The differences in stats are not due to random sampling or chance. Legendary Pokémon are mathematically designed to have higher base stats.
* **The "Total" Gap:** The 'Total' feature shows the most extreme significance ($p \approx 9.36 \times 10^{-47}$), confirming that the cumulative power gap is overwhelming.
* **Consistency:** The results are consistent across offensive (Attack/Sp. Atk), defensive (Defense/Sp. Def), and utility (HP/Speed) metrics.

| Feature | P-Value | Statistical Decision | Practical Conclusion |
| :--- | :--- | :--- | :--- |
| **All Features** | $< 0.05$ | Reject Null Hypothesis | Significantly Different |
| **Total Stat** | $\approx 0$ | Strong Rejection | Absolute Superiority |

#### Next, conduct t-test for Generation 1 and Generation 2 pokemons.

In [11]:
# Your code here
# Divide the dataset into Legendary and Non-Legendary groups
gen1_pokemon = pokemon[pokemon['Generation'] == 1]
gen2_pokemon = pokemon[pokemon['Generation'] == 2]

#Conduct the t-test for all features using function t_test_features
pokemon_ttest_results = t_test_features(gen1_pokemon, gen2_pokemon)

print("Generation 1 vs. Generation 2 Pokemon t-test results:")
for feature, p_value in pokemon_ttest_results.items():
    print(f"{feature:10}: {p_value}")

Generation 1 vs. Generation 2 Pokemon t-test results:
HP        : 0.14551697834219646
Attack    : 0.2472195896721772
Defense   : 0.5677711011725424
Sp. Atk   : 0.12332165977104394
Sp. Def   : 0.18829872292645747
Speed     : 0.0023926593731213447
Total     : 0.5631377907941677


#### What conclusions can you make?

In [ ]:
# Your comment here

### Conclusion: Generation 1 vs. Generation 2 Analysis

Based on the p-values from the independent t-tests, we can draw the following conclusions:

#### 1. Statistical Significance (The Speed Gap)
* **Speed is the only significant difference:** The p-value for **Speed (0.0024)** is the only one below our threshold of **0.05**. This means we **reject the null hypothesis** only for Speed. 
* **Conclusion:** Generation 1 Pokémon are significantly different (likely faster on average) than Generation 2 Pokémon.

#### 2. Non-Significant Differences (The Similarity)
* **Most stats are similar:** For HP, Attack, Defense, Sp. Atk, Sp. Def, and Total, the **p-values are all $> 0.05$** (ranging from 0.12 to 0.56).
* **Decision:** we **fail to reject the null hypothesis ($H_0$)** for these features. 
* **Conclusion:** There is no statistically significant evidence to suggest that Gen 1 and Gen 2 differ in their power, bulk, or overall stats.

#### 3. Key Takeaway
While Game Freak introduced many new Pokémon in the second generation, they maintained a very similar "power level" (Total) to the first generation. The only major shift was in the **Speed** meta, where the two generations show a clear statistical divergence.

| Feature | P-Value | Statistical Decision | Practical Conclusion |
| :--- | :--- | :--- | :--- |
| **Speed** | **0.0024** | **Reject $H_0$** | **Significantly Different** |
| **Total** | 0.5631 | Fail to Reject $H_0$ | No significant difference |
| **All others** | $> 0.05$ | Fail to Reject $H_0$ | No significant difference |

#### Compare pokemons who have single type vs those having two types.

In [20]:
# Your code here
# Divide the dataset into Generation 1 and Generation 2 groups
from numpy import nan
from sqlalchemy import null


only_one_type_pokemon = pokemon[pokemon['Type 2'].isna()]
two_types_pokemon = pokemon[pokemon['Type 2'].notna()]

#Conduct the t-test for all features using function t_test_features
pokemon_ttest_results = t_test_features(only_one_type_pokemon, two_types_pokemon)

print("One type vs. Two types Pokemon t-test results:")
for feature, p_value in pokemon_ttest_results.items():
    print(f"{feature:10}: {p_value}")

One type vs. Two types Pokemon t-test results:
HP        : 0.11314389855379418
Attack    : 0.00014932578145948315
Defense   : 2.7978540411514706e-08
Sp. Atk   : 0.00013876216585667923
Sp. Def   : 0.00010730610934512791
Speed     : 0.024217032818190903
Total     : 1.1157056505230023e-07


#### What conclusions can you make?

In [ ]:
# Your comment here

### Conclusion: Single-Type vs. Dual-Type Pokémon Analysis

Based on the independent t-test results, we can draw the following statistical conclusions:

#### 1. Widespread Statistical Significance
* **Almost all p-values are below 0.05:** With the sole exception of **HP (0.1131)**, every other feature shows a statistically significant difference between the two groups.
* **The "Total" Factor:** The p-value for the **Total** stat is extremely low ($1.11 \times 10^{-7}$). Therefore, we **strictly reject the null hypothesis ($H_0$)** for the overall power level.

#### 2. Key Insights
* **Inherent Power Advantage:** Dual-type Pokémon are not just tactically versatile; they are mathematically stronger. They possess significantly higher base stats in nearly every category compared to single-type Pokémon.
* **Offense and Defense:** The p-values for Attack, Defense, Sp. Atk, and Sp. Def are all remarkably low (all $< 0.0002$). This suggests that dual-type Pokémon are designed to be both more robust and more lethal.
* **The Speed Metric:** While significant ($p=0.024$), Speed is the closest to the 0.05 threshold. This indicates that while dual-types are generally faster, the gap is not as massive as it is for raw power or defensive capabilities.

#### 3. Strategic Reflection
The lack of significance in **HP** ($p=0.11$) suggests that adding a second type does not necessarily grant more "health" or vitality. Instead, the "evolutionary" advantage of a second type manifests as increased combat efficiency (**Attack/Sp. Atk**) and resilience (**Defense/Sp. Def**).

| Feature | P-Value | Statistical Decision | Practical Conclusion |
| :--- | :--- | :--- | :--- |
| **Total** | **< 0.000001** | **Reject $H_0$** | **Massive Difference** |
| **HP** | 0.1455 | Fail to Reject $H_0$ | No significant difference |
| **Offensive Stats**| **< 0.0002** | **Reject $H_0$** | **Significantly Stronger** |
| **Speed** | **0.0242** | **Reject $H_0$** | **Significantly Faster** |

#### Now, we want to compare whether there are significant differences of `Attack` vs `Defense`  and  `Sp. Atk` vs `Sp. Def` of all pokemons. Please write your code below.

*Hint: are you comparing different populations or the same population?*

In [25]:
# Your code here
# Same population we use ttest_rel instead of ttest_ind
from scipy.stats import ttest_rel

# Attack vs Defense
t_stat_att_def, p_val_att_def = ttest_rel(pokemon['Attack'], pokemon['Defense'])

#  Atk vs Sp. Def
t_stat_sp, p_val_sp = ttest_rel(pokemon['Sp. Atk'], pokemon['Sp. Def'])

print(f"Attack vs Defense p-value: {p_val_att_def}")
print(f"Sp. Atk vs Sp. Def p-value: {p_val_sp}")

Attack vs Defense p-value: 1.7140303479358582e-05
Sp. Atk vs Sp. Def p-value: 0.3933685997548118


#### What conclusions can you make?

In [ ]:
# Your comment here

### Conclusion: Internal Attribute Comparison (Paired T-Test)

By analyzing the same population (all Pokémon) and comparing their internal stats, we found:

#### 1. Physical Combat (Attack vs. Defense)
* **Result:** Statistically Significant ($p = 1.71 \times 10^{-5}$).
* **Insight:** There is a clear imbalance in physical attributes. Pokémon are not designed to have equal Attack and Defense levels on average. The "Physical Meta" is biased.

#### 2. Special Combat (Sp. Atk vs. Sp. Def)
* **Result:** Not Significant ($p = 0.3933$).
* **Insight:** We fail to reject the null hypothesis. The Special stats are remarkably well-balanced across the Pokémon population, suggesting a more symmetrical design in special attributes.

| Comparison | P-Value | Decision | Conclusion |
| :--- | :--- | :--- | :--- |
| **Attack vs Defense** | **0.000017** | **Reject $H_0$** | **Significantly Different** |
| **Sp. Atk vs Sp. Def** | 0.3933 | Fail to Reject $H_0$ | No significant difference |

#### Extra Analisys
- Having a significant difference between Attack and Defense makes sense, giving the game a random and interesting component. This difference creates a variety of encounters and outcomes, leading the player to use different Pokémon, not just one. They have to search for and defeat them, making the game very interesting and unpredictable. This prevents the player from getting bored quickly and ensures they can always find a Pokémon that defeats them, forcing them to search for new Pokémon with different stats.